numpy/core/setup_common.py

```diff
@@ -120,20 +120,38 @@ def set_sig(sig):
             set_sig(line)
 
 # Mandatory functions: if not found, fail the build
-MANDATORY_FUNCS = ["sin", "cos", "tan", "sinh", "cosh", "tanh", "fabs",
-        "floor", "ceil", "sqrt", "log10", "log", "exp", "asin",
-        "acos", "atan", "fmod", 'modf', 'frexp', 'ldexp']
-
-# Standard functions which may not be available and for which we have a
-# replacement implementation. Note that some of these are C99 functions.
-OPTIONAL_STDFUNCS = ["expm1", "log1p", "acosh", "asinh", "atanh",
-        "rint", "trunc", "exp2", "log2", "hypot", "atan2", "pow",
-        "copysign", "nextafter", "strtoll", "strtoull", "cbrt"]
+MANDATORY_FUNCS = [
+    "sin", "cos", "tan", "sinh", "cosh", "tanh", "fabs",
+    "floor", "ceil", "sqrt", "log10", "log", "exp", "asin",
+    "acos", "atan", "fmod", 'modf', 'frexp', 'ldexp',
+    "expm1", "log1p", "acosh", "asinh", "atanh",
+    "rint", "trunc", "exp2", 
+    "copysign", "nextafter", "strtoll", "strtoull", "cbrt",
+    "log2", "pow", "hypot", "atan2",
+]
 
 OPTIONAL_LOCALE_FUNCS = ["strtold_l"]
 OPTIONAL_FILE_FUNCS = ["ftello", "fseeko", "fallocate"]
 OPTIONAL_MISC_FUNCS = ["backtrace", "madvise"]
 
+# variable attributes tested via "int %s a" % attribute
+OPTIONAL_VARIABLE_ATTRIBUTES = ["__thread", "__declspec(thread)"]
+
+# Subset of OPTIONAL_*_FUNCS which may already have HAVE_* defined by Python.h
+OPTIONAL_FUNCS_MAYBE = [
+    "ftello", "fseeko"
+    ]
+
+# TODO: make these mandatory and use BLOCK_ macros rather than HAVE_ macros
+# in npy_config.h and throughout the code
```

Comment:
This comment seems our of date now as `npy_config.h` already uses the BLOCK_ variants.

Reply:
The "make these mandatory" part is also not yet directly possible, as MSVC does not support them yet at all. They do support C99 complex through their own types though, which should make it [possible](https://github.com/numpy/numpy/issues/20880#issuecomment-1210456286) to achieve effectively the same outcome (well, aside from having to carry shim types in numpy that switch accordingly)

Reply:
I think the shim part is already done, since cexp and cpow are not blocked for MSVC. Like glibc<2.18, many are blocked on MSVC because of accuracy problems. In any case, I will remove the comment.